### PR TITLE
safesocket: add connect retry loop to wait for tailscaled

### PR DESCRIPTION
**Pull request by bradfitz**
Updates #2708

### CODE REVIEW

File: safesocket/safesocket.go [hunk 1]
```diff
@@ -29,9 +33,54 @@ func ConnCloseWrite(c net.Conn) error {
 	return c.(closeable).CloseWrite()
 }
 
+var processStartTime = time.Now()
+
+// tailscaledStillStarting reports whether tailscaled is probably
```
Comment (maisem) [inline safesocket/safesocket.go:38]: This adds tailscale specific code in an otherwise generic library.

Comment (bradfitz) [reply safesocket/safesocket.go:38]: Do you propose we create `tailscale.com/safesocket/tailscalesafesocket` ?

Comment (bradfitz) [reply safesocket/safesocket.go:38]: (That was a joke.)

Look at `safesocket_darwin.go` to convince yourself that safesocket is Tailscale-specific already. :)

**Comment by dsnet** [comment]
What's the rationale for this approach, which incurs a new dependency over getting `tailscaled` to start listening on a socket ASAP, and then use that channel to further communicate whether it is ready or not?

**Comment by bradfitz** [comment]
> What's the rationale for this approach, which incurs a new dependency, 

We already depend on this package.

> over getting tailscaled to start listening on a socket ASAP, 

It *already does* start listening on the socket ASAP.

> and then use that channel to further communicate whether it is ready or not?

Because that's racey. That's where we're at today. A "tailscaled& tailscale" could still schedule the tailscale part to run first.

File: safesocket/safesocket.go (hunk 1 above)
Comment (maisem) [reply safesocket/safesocket.go:38]: Lol well ok then

File: safesocket/safesocket_ps.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+		}
+		for _, proc := range procs {
+			name := proc.Executable()
+			const tailscaled = "tailscaled"
+			if len(name) < len(tailscaled) {
+				continue
+			}
+			// Do case insensitive comparison for Windows,
+			// notably, and ignore any ".exe" suffix.
+			if strings.EqualFold(name[:len(tailscaled)], tailscaled) {
```
Comment (dsnet) [inline safesocket/safesocket_ps.go:30]: This only works for case insensitivity in ASCII, which I assume is sufficient? For non-ASCII, `name[:len(tailscaled)]` will truncate the string if there are multi-byte Unicode characters.

**Review by dsnet — COMMENTED** [review]
> Because that's racey. That's where we're at today. A "tailscaled& tailscale" could still schedule the tailscale part to run first.

The new logic is also still racy. Hard-coding things like <2s and >5s suggests that we're making assumptions about what a good startup looks like. That said, I guess it's better than before.

**Review by dsnet — APPROVED** [review]

File: safesocket/safesocket_ps.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build linux || windows || darwin || freebsd
+// +build linux windows darwin freebsd
```
Comment (maisem) [inline safesocket/safesocket_ps.go:6]: openbsd?

**Review by maisem — APPROVED** [review]

File: safesocket/safesocket_ps.go (hunk 3 above)
Comment (bradfitz) [reply safesocket/safesocket_ps.go:6]: The OpenBSD build failure is why I pulled this out to its own file in PS2. In the first version I had it in a portable file and discovered that the ps package doesn't support OpenBSD. So oh well.

File: safesocket/safesocket_ps.go (hunk 2 above)
Comment (bradfitz) [reply safesocket/safesocket_ps.go:30]: yup, sufficient. The second arg is ASCII anyway.

**Pull request merged**